File: Guia3/ej9/Triangle.java
package Guias.Guia3.ej9;
import java.awt.Point;

public class Triangle extends Figure{
    private final Point p1;
    private final Point p2;
    private final Point p3;

    public Triangle(Point p1, Point p2, Point p3) {
        this.p1 = p1;
        this.p2 = p2;
        this.p3 = p3;
    }

    @Override
    public double perimeter() {
        return side1() + side2() + side3();
    }

    public double area(){
        double semiperimeter = perimeter() / 2;
        return Math.sqrt(semiperimeter * (semiperimeter - side1()) * (semiperimeter - side2()) * (semiperimeter - side3()));
    }
    private double side1(){
        return p1.distance(p2);
    }
    private double side2(){
        return p2.distance(p3);
    }
    private double side3(){
        return p3.distance(p1);
    }

    @Override
    public String toString() {
        return "Triángulo [ %s , %s , %s ]".formatted(p1, p2, p3);
    }

}

File: POO-Ruby/Guia8/ej9/triangle.rb
# frozen_string_literal: true

class Triangle < Figure

  def initialize(p1, p2, p3)
    @p1=p1
    @p2=p2
    @p3=p3
  end

  def perimeter
    segment1_2 + segment2_3 + segment3_1
  end

  def area
    semiperimiter = (segment1_2 + segment2_3 + segment3_1) / 2
    Math.sqrt(semiperimiter * (semiperimiter - segment1_2) * (semiperimiter - segment2_3) * (semiperimiter - segment3_1))
  end

  def to_s
    "Triángulo [ #{@p1} , #{@p2} , #{@p3} ]"
  end

  def ==(other)
    false unless other.is_a? Triangle
    @p1==other.p1 and @p2==p2 and @p3==p3
  end

  protected
  attr_reader :p1, :p2, :p3

  private
  def segment1_2
    @p1.distance @p2
  end
  def segment2_3
    @p2.distance @p3
  end
  def segment3_1
    @p3.distance @p1
  end
end
